### ISSUE TITLE
Ingress for Optmize, Identity 

### ISSUE THREAD

**Issue by widdix123** [camunda/camunda-platform-helm#334]
Hi,

Change the ingress settings to true, adding the redirectURL to point to the ingress still it references to localhost and http
Please suggest what needs to be done 

https://camunda-operate.xxxx.xxxx/auth/realms/camunda-platform/protocol/openid-connect/auth?client_id=operate&redirect_uri=http%3A%2F%2Fcamunda-operate.xxxx.xxxx%2Fidentity-callback&response_type=code&scope=openid+email&state=

**Comment by widdix123** [comment]
Is there a documentation on how we can configure an ingress  ? I have set the ingress but nether keycloak, tasklist, identity or optmizie works

Some are redirect to http , or to localhost ports .

I dont see any configuration parameters for these where i could configure .

This is a blocker for us to have it deployed

**Comment by widdix123** [comment]
@Zelldon - It is better than before with below issues 

- Admin page to keycloak is not accessible yet . It has the below error in the network tab under Inspect 

```
Mixed Content: The page at 'https://camunda-keycloak.xxxx.dev/auth/admin/master/console/' was loaded over HTTPS, but requested an insecure script 'http://camunda-keycloak.xxxx.dev/auth/js/keycloak.js?version=450u6'. This request has been blocked; the content must be served over HTTPS.
```
-  the identity ingress redirects to the application but it throws this error  because the url `http:// `

URL after redirection: http://camunda-keycloak.xxx.xxx/auth/realms/camunda-platform/login-actions/authenticate?session_code=xxxxxxxxxx&execution=917fcca3-56df-443a-91b7-xxxxxxx&client_id=camunda-identity&tab_id=GBKn-LTPaus 

<img width="1179" alt="image" src="https://user-images.githubusercontent.com/38529736/174583044-8a721d54-b4f9-40d0-a705-1b91f4d70f61.png">

Only by clicking "Send Anyway" it redirects to the identity page

- Same issue with optimize and operate sites too 

Can you please suggest ?

**Comment by widdix123** [comment]
@Zelldon  - Any suggestions on the above comment ?

**Comment by Zelldon** [comment]
Would be helpful if you share your values file so we can see what you have configured. 

But in general it sounds like you want to enable https with your ingress, to get it work you also have to set `KEYCLOAK_PROXY_ADDRESS_FORWARDING=true` as environment variable on the keycloak subchart. This is an known issue with keycloak see https://stackoverflow.com/questions/32084604/why-is-keycloak-removing-the-ssl-in-the-redirect-uri

Example cfg:

```yaml
  identity:
    keycloak:
      extraEnvVars:
        - name: KEYCLOAK_PROXY_ADDRESS_FORWARDING
          value: "true"
```

**Comment by widdix123** [comment]
@Zelldon -

Attached values.yaml . Please note, i have used the same provided by you and updated my details

[camunda-values-updated.yaml.txt](https://github.com/camunda/camunda-platform-helm/files/8946860/camunda-values-updated.yaml.txt)

Also, Ingress is hosted using  nginx-ingress controllers with the below values.yaml 

helm install xxxx-ingress --values values.yaml ingress-nginx/ingress-nginx --namespace nginx-ingress

```
controller:
  # publishService required to Allow ELB Alias for DNS registration w/ external-dns
  publishService:
    enabled: true
  tcp:
    configNameSpace: $(POD_NAMESPACE)/tcp-services
  udp:
    configNameSpace: $(POD_NAMESPACE)/udp-services
  service:
    # AWS Annotations for LoadBalaner with Certificate ARN
    annotations:
      service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "arn:aws:acm:eu-west-3:xxxxxxx:certificate/xxxxxxxx-2441-4fbf-bd12-103219944c1e"
      service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "http"
      service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
      service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
      service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: "ELBSecurityPolicy-TLS-1-2-2017-01"
    # TLS (https) terminated at ELB, so internal endpoint is 'http'
    targetPorts:
      https: http
```

**Comment by widdix123** [comment]
@Zelldon  - What type of ingress controller are you using for above deployment ?

**Comment by widdix123** [comment]
@Zelldon  -

1. Is it expected that tasklist is disabled? - yes but we can enable it 
2. You use everywhere http in your URI is this expected? - I saw that you were using but i tried changing to https same issue

**Comment by widdix123** [comment]
@Zelldon  - That's correct ... I have a test site that you could check for issue.

I tried to do `https` too and it has the  same issue

Do you have similar issue with https. ?

For the `nginx-ingress`, the below includes ACM, http , https etc  

```
controller:
  # publishService required to Allow ELB Alias for DNS registration w/ external-dns
  publishService:
    enabled: true
  tcp:
    configNameSpace: $(POD_NAMESPACE)/tcp-services
  udp:
    configNameSpace: $(POD_NAMESPACE)/udp-services
  service:
    # AWS Annotations for LoadBalaner with Certificate ARN
    annotations:
      service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "arn:aws:acm:eu-west-3:478653635976:certificate/a7246943-2441-4fbf-bd12-103219944c1e"
      service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "http"
      service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
      service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
      service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: "ELBSecurityPolicy-TLS-1-2-2017-01"
    # TLS (https) terminated at ELB, so internal endpoint is 'http'
    targetPorts:
      https: http
```

**Comment by Zelldon** [comment]
Sorry I had no time yet to check this. 

The values file I sent to you was used by a colleague of mine. I can try to investigate the https and ingress topic, if I find the time, but just to answer your question we have run our experiments against GKE. The helm charts in general are also tested only against GKE. I hope we will have time soon to validate them against other cloud providers.

**Comment by widdix123** [comment]
@Zelldon  - I set the tls: true and tried  but still the same error

this is my latest values.yaml. Please check 

[camunda-values-updated.txt](https://github.com/camunda/camunda-platform-helm/files/8973142/camunda-values-updated.txt)

**Comment by widdix123** [comment]
@Zelldon  - Infact now I am getting this after setting the `tls` to true for operate and optimize 

<img width="1221" alt="image" src="https://user-images.githubusercontent.com/38529736/175464645-d955f8e7-d583-456d-976d-6419af9f049e.png">

**Comment by widdix123** [comment]
@Zelldon  - yes everytime i do a clean installation . Helm delete and pvc delete each time for the complete namespace

**Comment by widdix123** [comment]
URL looks like : 

https://camunda-keycloak.test.dev/auth/realms/camunda-platform/protocol/openid-connect/auth?client_id=operate&redirect_uri=http%3A%2F%2Fcamunda-operate.test.dev%2Fidentity-callback&response_type=code&scope=openid+email&state=

**Comment by widdix123** [comment]
@Zelldon  - Any suggestion on how can i fix the keycloak issue ?

**Comment by widdix123** [comment]
i found the fix for this . We need to add the following in the extraEnvVars for keycloak

Both `KEYCLOAK_PROXY_ADDRESS_FORWARDING` and `KEYCLOAK_EXTRA_ARGS` are important as in EKS the `KEYCLOAK_FRONTEND_URL` does not get updated to the keycloak configuration

```
  # Keycloak configuration, for the keycloak dependency chart which is used by identity
  keycloak: # see: http://github.com/bitnami/charts/tree/master/bitnami/keycloak#parameters
    ingress: # see: http://www.keycloak.org/server/reverseproxy
      enabled: true
      ingressClassName: nginx
      hostname: "keycloak.test.dev"
    extraEnvVars: # see: http://github.com/bitnami/bitnami-docker-keycloak#configuration
      - name: KEYCLOAK_PROXY_ADDRESS_FORWARDING
        value: "true"
      - name: KEYCLOAK_EXTRA_ARGS
        value: "-Dkeycloak.frontendUrl=https://keycloak.test.dev/auth"
```
   This has resolved all the issues related to https, page not secure

**Comment by widdix123** [comment]
@Zelldon @oleschoenburg - There is another issue with the zeebe gateway and it is not working with port 443

Can you please suggest how can we use the zeebe gateway on port 443 for gprc protocol

```
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: camunda-zeebe-gateway
  namespace: camunda
  annotations:
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/backend-protocol: "GRPC"
spec:
  ingressClassName: nginx
  rules:
    # route by virtualhost - externaldns registers this record
  - host: camunda-zeebe-gateway.test.dev
    http:
      paths:
      - backend:
          serviceName: camunda-zeebe-gateway
          # connect to service using http port
          servicePort: 26500
        # default path put here explicitly for illustrative purposes
        path: /
```

**Comment by oleschoenburg** [comment]
Hey @widdix123,

Usually we recommend to terminate TLS at the ingress (see for example here: https://kubernetes.github.io/ingress-nginx/examples/tls-termination/, essentially you'd just have to add a tls section) but in [this comment](https://github.com/camunda/camunda-platform-helm/issues/334#issuecomment-1161388164) it appears that you terminate TLS on the AWS load balancer already.

In that case nginx-ingress [suggest](https://kubernetes.github.io/ingress-nginx/user-guide/tls/#server-side-https-enforcement-through-redirect)  to force the https redirect by setting `nginx.ingress.kubernetes.io/force-ssl-redirect: "true"`.

> There is another issue with the zeebe gateway and it is not working with port 443

Could you expand a little bit what exactly doesn't work? Is it that the client can't connect? Any logs or error messages would be useful for me to understand the situation a bit better. Thanks!

**Comment by oleschoenburg** [comment]
But in the values.yml from your [previous comment](https://github.com/camunda/camunda-platform-helm/issues/334#issuecomment-1165184509) it looks like you are trying to terminate TLS at the ingress and not at the load balancer anymore. I that case you could just try and add the TLS section to the ingress as I've mentioned above:

```diff
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: camunda-zeebe-gateway
  namespace: camunda
  annotations:
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/backend-protocol: "GRPC"
spec:
  ingressClassName: nginx
  rules:
    # route by virtualhost - externaldns registers this record
  - host: camunda-zeebe-gateway.test.dev
    http:
      paths:
      - backend:
          serviceName: camunda-zeebe-gateway
          # connect to service using http port
          servicePort: 26500
        # default path put here explicitly for illustrative purposes
        path: /
+ - tls:
+  - hosts
+    - camunda-zeebe-gateway.test.dev
```

**Comment by oleschoenburg** [comment]
Since it is difficult for me to replicate that setup (TLS terminated by an AWS loadbalancer), could you still explain what exactly isn't working? I.e. logs or error messages when trying to connect to the gateway.

**Comment by widdix123** [comment]
<img width="301" alt="image" src="https://user-images.githubusercontent.com/38529736/176842847-11f12b7e-39be-4e67-accc-1c2c3b5b650e.png">

@oleschoenburg  - This is what we get in the workflow moduler

**Comment by widdix123** [comment]
@Zelldon  @oleschoenburg  - Any suggestion on above ?

**Comment by oleschoenburg** [comment]
So if I understand correctly:

The gateway is running on the default port, 26500. You defined a loadbalancer that listens on port 443, terminates TLS and forwards to the gateway port 26500. In this setup the modeler can't connect on port 443. When you add a different loadbalancer that listens on port 26500 and forwards to the gateway on port 26500 everything works and the modeler can connect. 

Is that accurate? To me that'd indicate that the AWS loadbalancer setup is not correct. Since I don't know how to get debug logs from the modeler, could you try connecting via [zbctl](https://docs.camunda.io/docs/apis-clients/cli-client/)? Maybe that'll print out some details on why the connection fails.

**Comment by widdix123** [comment]
@oleschoenburg  - Yes that is correct . The http loadbalancer on port 26500 works fine and connect to the modeler

Can you share me the zbctl command that you would want me to execute ?

**Comment by widdix123** [comment]
@oleschoenburg 

for the loadbalancer: (GRPC_GO_LOG_SEVERITY_LEVEL=info zbctl status --insecure)  results in

```
2022/07/05 16:06:31 INFO: [core] original dial target is: "xxxxxxx-956261995.eu-west-3.elb.amazonaws.com:26500"
2022/07/05 16:06:31 INFO: [core] parsed dial target is: {xxxxxxx-956261995.eu-west-3.elb.amazonaws.com Authority: Endpoint:26500 URL:{xxxxxxx-956261995.eu-west-3.elb.amazonaws.com Opaque:26500 User: Host: Path: RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}
2022/07/05 16:06:31 INFO: [core] fallback to scheme "passthrough"

2022/07/05 16:06:31 INFO: [core] parsed dial target is: {Scheme:passthrough Authority: Endpoint:xxxxxxx-956261995.eu-west-3.elb.amazonaws.com:26500 URL:{Scheme:passthrough Opaque: User: Host: Path:/xxxxxxx-956261995.eu-west-3.elb.amazonaws.com:26500 RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}
2022/07/05 16:06:31 INFO: [core] Channel authority set to "xxxxx-956261995.eu-west-3.elb.amazonaws.com:26500"
2022/07/05 16:06:31 INFO: [core] ccResolverWrapper: sending update to cc: {[{xxxxxx-956261995.eu-west-3.elb.amazonaws.com:26500  <nil> <nil> 0 <nil>}] <nil> <nil>}
2022/07/05 16:06:31 INFO: [core] ClientConn switching balancer to "pick_first"
2022/07/05 16:06:31 INFO: [core] Channel switches to new LB policy "pick_first"
2022/07/05 16:06:31 INFO: [core] blockingPicker: the picked transport is not ready, loop back to repick
2022/07/05 16:06:31 INFO: [core] Subchannel Connectivity change to CONNECTING
2022/07/05 16:06:31 INFO: [core] Subchannel picks a new address "xxxxxx-956261995.eu-west-3.elb.amazonaws.com:26500" to connect
2022/07/05 16:06:31 INFO: [core] Channel Connectivity change to CONNECTING

2022/07/05 16:06:36 INFO: [core] CPU time info is unavailable on non-linux environments.

2022/07/05 16:06:36 INFO: [core] Subchannel Connectivity change to READY

2022/07/05 16:06:36 INFO: [core] Channel Connectivity change to READY

Cluster size: 1
Partitions count: 1
Replication factor: 1
Gateway version: 8.0.2
Brokers:
  Broker 0 - camunda-zeebe-0.camunda-zeebe.camunda.svc:26501
    Version: 8.0.2
    Partition 1 : Leader, Healthy

2022/07/05 16:06:46 INFO: [core] Channel Connectivity change to SHUTDOWN
2022/07/05 16:06:46 INFO: [core] Subchannel Connectivity change to SHUTDOWN
```

For the https ingress url :

```
2022/07/05 16:05:06 INFO: [core] original dial target is: "camunda-zeebe-gateway.test.dev:443"

2022/07/05 16:05:06 INFO: [core] parsed dial target is: {Scheme:camunda-zeebe-gateway.test.dev Authority: Endpoint:443 URL:{Scheme:camunda-zeebe-gateway.test.dev Opaque:443 User: Host: Path: RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}
2022/07/05 16:05:06 INFO: [core] fallback to scheme "passthrough"
2022/07/05 16:05:06 INFO: [core] parsed dial target is: {Scheme:passthrough Authority: Endpoint:camunda-zeebe-gateway.test.dev:443 URL:{Scheme:passthrough Opaque: User: Host: Path:/camunda-zeebe-gateway.test.dev:443 RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}
2022/07/05 16:05:06 INFO: [core] Channel authority set to "camunda-zeebe-gateway.dixio.dev:443"
2022/07/05 16:05:06 INFO: [core] ccResolverWrapper: sending update to cc: {[{camunda-zeebe-gateway.test.dev:443  <nil> <nil> 0 <nil>}] <nil> <nil>}
2022/07/05 16:05:06 INFO: [core] ClientConn switching balancer to "pick_first"
2022/07/05 16:05:06 INFO: [core] Channel switches to new LB policy "pick_first"
2022/07/05 16:05:06 INFO: [core] Subchannel Connectivity change to CONNECTING
2022/07/05 16:05:06 INFO: [core] Subchannel picks a new address "camunda-zeebe-gateway.test.dev:443" to connect
2022/07/05 16:05:06 INFO: [core] blockingPicker: the picked transport is not ready, loop back to repick
2022/07/05 16:05:06 INFO: [core] Channel Connectivity change to CONNECTING

2022/07/05 16:05:11 INFO: [core] CPU time info is unavailable on non-linux environments.

2022/07/05 16:05:12 INFO: [core] Subchannel Connectivity change to TRANSIENT_FAILURE

2022/07/05 16:05:12 INFO: [core] Channel Connectivity change to TRANSIENT_FAILURE
Error: rpc error: code = Unavailable desc = connection closed before server preface received

```

**Comment by widdix123** [comment]
@oleschoenburg  - Can you please take a look and suggest what could be the issue here?

**Comment by oleschoenburg** [comment]
I was able to verify that it is possible to access the zeebe gateway via an ingress that terminates TLS. 
Here's what I did:

First, install camunda-platfrom, nginx-ingress, cert-manager and the ingress for the zeebe gateway:
```

$ helm upgrade --install ingress-nginx ingress-nginx   --repo https://kubernetes.github.io/ingress-nginx   --namespace ingress-nginx --create-namespace
$ kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.8.2/cert-manager.yaml
$ helm install camunda-platform camunda/camunda-platform -f values.yml
$ kubectl apply -f zeebe-ingress.yml
```

values.yml:

```yaml
global:
  image:
    tag: "8.0.3"
  identity:
    auth:
      enabled: false
elasticsearch:
  replicas: 1
operate:
  env:
    - name: SERVER_SERVLET_CONTEXT_PATH
      value: /operate
  ingress:
    enabled: true
    host: xxxx.ip.linodeusercontent.com
    path: /operate
    pathType: Prefix
    annotations:
      cert-manager.io/issuer: "letsencrypt-staging"
    tls:
      enabled: true
      secretName: operate-cert
tasklist:
  ingress:
    enabled: true
    host: xxxx.ip.linodeusercontent.com
    path: /tasklist
    pathType: Prefix
optimize:
  ingress:
    enabled: true
    host: xxxx.ip.linodeusercontent.com
    path: /optimize
    pathType: Prefix
```

zeebe-ingress.yml:
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: zeebe-gateway
  annotations:
    kubernetes.io/ingress.class: "nginx"    
    cert-manager.io/issuer: "letsencrypt-staging"
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/backend-protocol: "GRPC"
spec:
  tls:
  - hosts:
    - xxxx.ip.linodeusercontent.com
    secretName: zeebe-gateway-cert
  rules:
  - host: xxxx.ip.linodeusercontent.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: camunda-platform-zeebe-gateway
            port:
              number: 26500
```


After downloading the letsencrypt-staging cert and saving it as gateway.cert I was able to connect:
```
$ zbctl --address xxxx.ip.linodeusercontent.com:443 --certPath=gateway.cert status
Cluster size: 3
Partitions count: 3
Replication factor: 3
Gateway version: 8.0.3
Brokers:
  Broker 0 - camunda-platform-zeebe-0.camunda-platform-zeebe.default.svc:26501
    Version: 8.0.3
    Partition 1 : Leader, Healthy
    Partition 2 : Follower, Healthy
    Partition 3 : Follower, Healthy
  Broker 1 - camunda-platform-zeebe-1.camunda-platform-zeebe.default.svc:26501
    Version: 8.0.3
    Partition 1 : Follower, Healthy
    Partition 2 : Leader, Healthy
    Partition 3 : Leader, Healthy
  Broker 2 - camunda-platform-zeebe-2.camunda-platform-zeebe.default.svc:26501
    Version: 8.0.3
    Partition 1 : Follower, Healthy
    Partition 2 : Follower, Healthy
    Partition 3 : Follower, Healthy
```

If you are still having issues with the AWS loadbalancer I'd recommend double-checking this resource: https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/deploy-a-grpc-based-application-on-an-amazon-eks-cluster-and-access-it-with-an-application-load-balancer.html

Please be aware that we generally don't recommend to expose the zeebe gateway to the internet - the gateway does not use authentication so your zeebe cluster would be freely available to anyone. There is one community project that I know of that adds authentication via keycloak: https://github.com/camunda-community-hub/zeebe-keycloak-interceptor/

In summary: The webapps work as expected behind TLS, zeebe should too but the setup might be a bit more tricky because it's GRPC, not just regular HTTP. I'd suggest that you open a question at the forum if you really want to expose zeebe to the internet and are still having issues with TLS. Maybe there are other people around that have more experience with AWS that can help you.

Thanks for raising this issue anyway and for helping by posting logs etc. :+1:
